src/services/pcn-bridge/src/Bridge_dp_no_stp.c

```diff
@@ -0,0 +1,310 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bcc/helpers.h>
+#include <bcc/proto.h>
+
+#include <uapi/linux/bpf.h>
+#include <uapi/linux/filter.h>
+#include <uapi/linux/if_ether.h>
+#include <uapi/linux/if_packet.h>
+#include <uapi/linux/if_vlan.h>
+
+#include <uapi/linux/in.h>
+#include <uapi/linux/ip.h>
+#include <uapi/linux/pkt_cls.h>
+
+enum { BPDU_REASON = 1, BROADCAST_REASON };
+
+enum { PORT_MODE_ACCESS = 0, PORT_MODE_TRUNK };
+
+enum entry_type { STATIC = 0, DYNAMIC };
+
+#define VLAN_WILDCARD 0x00
+
+/*
+ * Allowed Vlans table is used to set and verify vlans associated to a port
+ * in access or trunk mode. The usage is provided in the following example.
+ */
+struct port_vlan_key {
+  u32 port;
+  u16 vlan;
+} __attribute__((packed));
+
+struct port_vlan_value {
+  u16 vlan;
+  u16 stp_state;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct port_vlan_key, struct port_vlan_value, port_vlan,
+          1024);
+/*
+ * port_vlan usage example:
+ * <port> <vlan> -> vlanid
+ *
+ * port 1 access vlan=10
+ * 1 VLAN_WILDCARD -> 10
+ *
+ * port 3 trunk vlan=30,40
+ * 3 30 -> 30
+ * 3 40 -> 40
+ *
+ * port 4 trunk vlan=all
+ * 4 VLAN_WILDCARD -> VLAN_WILDCARD
+ */
+
+/*
+ * Fowarding Table: VLAN + MAC Address to port association.  This table is
+ * filled
+ * in the learning phase and then is used in the forwarding phase to decide
+ * where to send a packet
+ */
+struct fwd_key {
+  u16 vlan;
+  __be64 mac;
+} __attribute__((packed));
+
+struct fwd_entry {
+  u64 timestamp;
+  u32 port;
+  enum entry_type type;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct fwd_key, struct fwd_entry, fwdtable, 1024);
+
+/*
+ * Port Table: Saves the state of the ports
+ * TODO: what is the best option in this case, hash vs array?
+ */
+struct port {
+  u16 mode;
+  u16 native_vlan;
+  bool native_vlan_enabled;
+};
+
+BPF_TABLE("hash", u32, struct port, ports, 256);
+
+struct eth_hdr {
+  __be64 dst : 48;
+  __be64 src : 48;
+  __be16 proto;
+} __attribute__((packed));
+
+static int handle_rx(struct CTXTYPE *ctx, struct pkt_metadata *md) {
+  void *data = (void *)(long)ctx->data;
+  void *data_end = (void *)(long)ctx->data_end;
+  struct eth_hdr *eth = data;
+
+  if (data + sizeof(*eth) > data_end)
+    goto DROP;
+
+  u32 in_ifc = md->in_port;
+
+  pcn_log(ctx, LOG_DEBUG, "in_ifc=%d proto=0x%x", in_ifc,
+          bpf_ntohs(eth->proto));
+  pcn_log(ctx, LOG_DEBUG, "mac src:%M dst:%M", eth->src, eth->dst);
+
+  u16 vlanid = 0;
+#ifdef POLYCUBE_XDP
```

Comment:
@mauriciovasquezbernal : can we define the above helpers to run both in case of XDP and TC? It looks cumbersome to me the necessity to use  #ifdef in the dataplane.

Reply:
Yes, we can.  I'll define `pcn_is_vlan_present` and `pcn_get_vlan_id` in TC so it is available in both cases.